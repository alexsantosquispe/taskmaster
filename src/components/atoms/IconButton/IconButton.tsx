import '@testing-library/jest-dom';

import { ALIGNMENT_TYPES, type AlignmentType } from '@/models/types';
import cn from 'clsx';
import type { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import { TooltipWrapper } from '../TooltipWrapper/TooltipWrapper';

interface IconButtonProps {
  ariaLabel: string;
  icon: ReactNode;
  onClick: () => void;
  isDisable?: boolean;
  isDefault?: boolean;
  tooltipMessage?: string;
  tooltipAlignment?: AlignmentType;
  className?: string;
}

export const IconButton = ({
  ariaLabel,
  onClick,
  icon,
  isDisable = false,
  isDefault = true,
  tooltipMessage = '',
  tooltipAlignment = ALIGNMENT_TYPES.TOP,
  className
}: IconButtonProps) => {
  return (
    <TooltipWrapper
      tooltipMessage={tooltipMessage}
      tooltipAlignment={tooltipAlignment}
      componentToRender={(setShowTooltip) => {
        return (
          <button
            type="button"
            aria-label={ariaLabel}
            onClick={() => {
              onClick();
              setShowTooltip(false);
            }}
            disabled={isDisable}
            className={twMerge(
              'text-primary/80 flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70',
              cn({
                'hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15':
                  !isDisable && isDefault,
                'hover:text-primary hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15 dark:hover:text-white':
                  !isDisable && !isDefault,
                'cursor-not-allowed bg-neutral-100 opacity-75 dark:bg-neutral-600':
                  isDisable
              }),
              className
            )}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
          >
            {icon}
          </button>
        );
      }}
    />
  );
};
