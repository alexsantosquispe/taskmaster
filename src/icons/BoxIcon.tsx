import { twMerge } from 'tailwind-merge';
import type { SvgIconProps } from '../models/types';

export const BoxIcon = ({ className }: SvgIconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      viewBox="0 0 16 16"
      fill="none"
      id="Box--Streamline-Majesticons"
      height={16}
      width={16}
      className={twMerge('size-6', className)}
    >
      <desc>{'\n    Box Streamline Icon: https://streamlinehq.com\n  '}</desc>
      <path
        fill="currentColor"
        fillRule="evenodd"
        d="M7.6466666666666665 1.4346666666666668a0.6666666666666666 0.6666666666666666 0 0 1 0.7066666666666667 0l4.602666666666666 2.876666666666667L8 7.226666666666667 3.0439999999999996 4.311333333333333l4.602666666666666 -2.876666666666667zM2.005333333333333 5.247333333333334A0.6673333333333332 0.6673333333333332 0 0 0 2 5.333333333333333v5.333333333333333a0.6666666666666666 0.6666666666666666 0 0 0 0.3133333333333333 0.5653333333333332L7.333333333333333 14.369333333333332v-5.9879999999999995L2.005333333333333 5.246666666666666zM8.666666666666666 14.369333333333332l5.02 -3.1373333333333333A0.6666666666666666 0.6666666666666666 0 0 0 14 10.666666666666666V5.333333333333333c0 -0.028666666666666663 -0.002 -0.057999999999999996 -0.005333333333333333 -0.086L8.666666666666666 8.380666666666666v5.988666666666667z"
        clipRule="evenodd"
        strokeWidth={0.6667}
      />
    </svg>
  );
};
