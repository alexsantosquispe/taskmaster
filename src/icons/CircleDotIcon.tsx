import type { SvgIconProps } from '../models/types';
import { twMerge } from 'tailwind-merge';

export const CircleDotIcon = ({ className }: SvgIconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={twMerge('size-5', className)}
    >
      <circle cx="12" cy="12" r="10" />
      <circle cx="12" cy="12" r="1" />
    </svg>
  );
};
