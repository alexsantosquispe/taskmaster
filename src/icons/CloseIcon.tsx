import { twMerge } from 'tailwind-merge';
import type { SvgIconProps } from '../models/types';

export const CloseIcon = ({ className }: SvgIconProps) => {
  return (
    <svg
      data-testid="close-icon"
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={twMerge('size-5', className)}
    >
      <path d="M18 6 6 18" />
      <path d="m6 6 12 12" />
    </svg>
  );
};
