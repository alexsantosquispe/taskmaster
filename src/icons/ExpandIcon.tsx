import { twMerge } from 'tailwind-merge';
import type { SvgIconProps } from '../models/types';

export const ExpandIcon = ({ className }: SvgIconProps) => {
  return (
    <svg
      data-testid="expand-icon"
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={twMerge('size-5', className)}
    >
      <rect width="18" height="18" x="3" y="3" rx="2" />
      <path d="M15 3v18" />
      <path d="m8 9 3 3-3 3" />
    </svg>
  );
};
