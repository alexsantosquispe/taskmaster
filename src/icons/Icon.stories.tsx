import * as Icons from './index';

import type { Meta } from '@storybook/react';
import { useState } from 'react';

export default {
  title: 'Icons'
} as Meta;

export const IconGallery = () => {
  const [filter, setFilter] = useState('');

  const filteredIcons = Object.entries(Icons).filter(([name]) =>
    name.toLowerCase().includes(filter.toLowerCase())
  );

  return (
    <div className="space-y-6 p-4">
      <div className="flex w-1/2 flex-col gap-y-1">
        <label>Icons Gallery</label>
        <input
          placeholder="Search icons..."
          value={filter}
          onChange={(e) => setFilter(e.target.value)}
          className="rounded-md border border-neutral-300 p-2"
        />
      </div>

      <div className="flex flex-wrap gap-4">
        {filteredIcons.map(([name, IconComponent]) => (
          <div
            key={name}
            className="flex h-[10rem] w-[10rem] flex-col items-center justify-center gap-4 rounded-lg border border-neutral-300"
          >
            <IconComponent className="size-6" />
            <pre className="text-primary text-xs font-medium">{name}</pre>
          </div>
        ))}
      </div>
    </div>
  );
};
