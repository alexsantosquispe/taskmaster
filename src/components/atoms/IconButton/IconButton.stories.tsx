import type { Meta, StoryObj } from '@storybook/react';

import { CirclePlusIcon } from '@/icons';
import { Wrapper } from '@/utils/testing/unitTest.util';
import { IconButton } from './IconButton';

const meta: Meta<typeof IconButton> = {
  title: 'Atoms/IconButton',
  component: IconButton,
  parameters: {
    layout: 'centered'
  }
};

export default meta;

type Story = StoryObj<typeof IconButton>;

export const Default: Story = () => {
  return (
    <Wrapper>
      <IconButton
        ariaLabel="Icon button"
        icon={<CirclePlusIcon />}
        onClick={() => {}}
        tooltipMessage="Tooltip message"
      />
    </Wrapper>
  );
};

Default.args = {};
