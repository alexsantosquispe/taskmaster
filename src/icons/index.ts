import { BoardIcon } from './BoardIcon';
import { BoxIcon } from './BoxIcon';
import { CalendarIcon } from './CalendarIcon';
import { ChartPieIcon } from './ChartPieIcon';
import { ChevronDownIcon } from './ChevronDownIcon';
import { ChevronUpIcon } from './ChevronUpIcon';
import { CircleCheckIcon } from './CircleCheckIcon';
import { CircleDashedIcon } from './CircleDashedIcon';
import { CircleDotDashedIcon } from './CircleDotDashedIcon';
import { CirclePlusIcon } from './CirclePlusIcon';
import { CollapseIcon } from './CollapseIcon';
import { EllipsisVerticalIcon } from './EllipsisVerticalIcon';
import { ExpandIcon } from './ExpandIcon';
import { InfoIcon } from './InfoIcon';
import { LaptopIcon } from './LaptopIcon';
import { LoadIcon } from './LoadIcon';
import { MoonIcon } from './MoonIcon';
import { ProjectIcon } from './ProjectIcon';
import { SettingsIcon } from './SettingsIcon';
import { SunIcon } from './SunIcon';
import { TasksListIcon } from './TasksListIcon';

export {
  BoardIcon,
  BoxIcon,
  CalendarIcon,
  ChartPieIcon,
  ChevronDownIcon,
  ChevronUpIcon,
  CircleCheckIcon,
  CircleDashedIcon,
  CircleDotDashedIcon,
  CirclePlusIcon,
  CollapseIcon,
  EllipsisVerticalIcon,
  ExpandIcon,
  InfoIcon,
  LaptopIcon,
  LoadIcon,
  MoonIcon,
  ProjectIcon,
  SettingsIcon,
  SunIcon,
  TasksListIcon
};
