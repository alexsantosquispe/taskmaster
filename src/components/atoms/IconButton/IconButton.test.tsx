import { fireEvent, render, screen } from '@testing-library/react';

import { CirclePlusIcon } from '@/icons';
import { IconButton } from './IconButton';

describe('IconButton', () => {
  const onClickMock = jest.fn();
  const props = {
    ariaLabel: 'Icon button',
    icon: <CirclePlusIcon />,
    onClick: onClickMock,
    tooltipMessage: 'Tooltip message'
  };

  describe('styles', () => {
    it('should render the component correctly', () => {
      let component = render(<IconButton {...props} />);
      expect(component).toMatchSnapshot();

      component = render(<IconButton {...props} tooltipMessage="tooltip" />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should check if the button is disabled', () => {
      render(<IconButton {...props} isDisable={true} />);

      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('should display/hide a tooltip when the mouse hovers over the button', () => {
      render(<IconButton {...props} />);

      const iconButton = screen.getByRole('button');

      expect(iconButton).toBeInTheDocument();

      fireEvent.mouseEnter(iconButton);

      expect(screen.getByText('Tooltip message')).toBeInTheDocument();

      fireEvent.mouseLeave(iconButton);

      expect(screen.queryByText('Tooltip message')).not.toBeInTheDocument();
    });

    it('should call the onClick function when click the button', () => {
      render(<IconButton {...props} />);

      const iconButton = screen.getByRole('button');

      expect(iconButton).toBeInTheDocument();

      fireEvent.click(iconButton);

      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
});
