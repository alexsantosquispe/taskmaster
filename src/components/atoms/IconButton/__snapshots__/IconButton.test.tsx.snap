// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`IconButton should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex items-center justify-center"
        data-testid="tooltip-wrapper"
      >
        <button
          aria-label="Icon button"
          class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15"
          type="button"
        >
          <svg
            class="size-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="M8 12h8"
            />
            <path
              d="M12 8v8"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex items-center justify-center"
      data-testid="tooltip-wrapper"
    >
      <button
        aria-label="Icon button"
        class="text-primary flex w-fit items-center justify-center rounded-lg p-2 transition-colors duration-150 ease-in-out dark:text-white/70 hover:text-accent dark:hover:text-accent-dark hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/15"
        type="button"
      >
        <svg
          class="size-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M8 12h8"
          />
          <path
            d="M12 8v8"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
